### DISCUSSION TITLE
JPA Storage Ignoring Cache Configuration / 20.0.3

### DISCUSSION THREAD

**Discussion by AKermodeBear** [keycloak/keycloak#16809] [Q&A]
Hoping someone can help me out; it may be a misunderstanding from the documentation I've been able to find.

I currently have a HA cluster running on top of MariaDB/Remote Infinispan, but I would like to replace MariaDB with CRDB. The problem I'm running into is this:

JPA needs to be enabled for Keycloak to properly talk to CockroachDB. However, when --storage=jpa is set, it seems to completely ignore cache and cache-config-file configurations. Even when configuring Keycloak to look for a cache configuration file that doesn't exist, it starts right up. So there seems to be no attempt at all to interact with the cache configuration file.

If I attempt to start Keycloak without --storage=jpa, then Keycloak seems to read the cache configuration file (and it properly complains if it isn't there / invalid, so I know it's at least being looked at), but it then attempts to execute some updates via Liquibase and it fails.

So I am wondering at this point if enabling JPA is intended to completely replace the Infinispan cache layer, I'm missing part of the documentation, or there's some non-obvious configuration issue that I've run into. 

I did notice that there's the ability to configure hotrod storage, and then section off various parts of Keycloak's data to either hotrod or the database, but that isn't the caching layer.

I've looked at the Cache documentation, a few scattered articles on Keycloak and CRDB, the RedHat 3-part series on setting something like this up, etc., but it seems I've missed something along the way.

Any help would be appreciated. Thanks!

**Comment by thuongvo19** [comment]
Did you have a solution for it @AKermodeBear?
I am facing the same issue here.
What I am doing as a workaround is
```
RemoteCache<String, AggregatedHealthStatus> cacheMap = this.cacheDataIspn.getRemoteCache(CACHE_KEY);
  if (Objects.isNull(cacheMap)) {
      try {
          Field field = DefaultHotRodConnectionProvider.class.getDeclaredField("remoteCacheManager");
          field.setAccessible(true);
          RemoteCacheManager remoteCacheManager = (RemoteCacheManager) field.get(this.cacheDataIspn);
          remoteCacheManager.getConfiguration().addRemoteCache(CACHE_KEY, 
                (builder) -> builder.configurationURI(getCacheConfigUri(CACHE_KEY)));
          cacheMap = this.cacheDataIspn.getRemoteCache(CACHE_KEY);
      } catch (NoSuchFieldException | IllegalAccessException e) {
          throw new RuntimeException(e);
      }
  }
```

**Reply by AKermodeBear** [reply]
Nope, I gave up trying to make it work since I didn't want to dig into the Keycloak internals too much. Just went forward without JPA and told Keycloak to connect to CRDB as PGSQL (since CRDB is protocol-compatible). There were a few minor snags getting the initial schema set up, but after that everything seems to be humming along just fine.

Haven't experimented with JPA on the 21.x release (not sure if they changed anything), but at this point I don't need to, so. If I end up returning to this I'll definitely let you know, though. Thanks!
